pkg/apis/clusterlink.net/v1alpha1/instance_types.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2023 The ClusterLink Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// StatusConditionType represents the status conditions type for ClusterLink components.
+type StatusConditionType string
+
+const (
+	// DeploymentReady means the component deployment is ready to use.
+	DeploymentReady StatusConditionType = "DeploymentReady"
+)
+
+// IngressType represents the ingress type  of the deployed ClusterLink.
+type IngressType string
+
+const (
+	// IngressTypeNone indicates that the deployment instance did not create any external ingress.
+	IngressTypeNone IngressType = "none"
+	// IngressTypeNodePort indicates that the deployment instance created an external ingress of type NodePort.
+	// The service will use nodePort:30443.
+	IngressTypeNodePort IngressType = "NodePort"
+	// IngressTypeLoadBalancer indicates that the deployment instance created an external ingress of type LoadBalancer.
+	IngressTypeLoadBalancer IngressType = "LoadBalancer"
+)
+
+// DataplaneType represents the dataplane type  of the deployed ClusterLink.
+type DataplaneType string
+
+const (
+	// GoType indicates that the dataplane type is Go-dataplane.
+	GoType DataplaneType = "go"
+	// EnvoyType indicates that the dataplane type is Envoy.
+	EnvoyType DataplaneType = "envoy"
+)
+
+// InstanceSpec defines the desired state of a ClusterLink instance.
+type InstanceSpec struct {
+	DataPlane DataPlaneSpec `json:"dataplane,omitempty"`
+	Ingress   IngressSpec   `json:"ingress,omitempty"`
+
+	// +kubebuilder:validation:Enum=trace;debug;info;warning;error;fatal
+	// +kubebuilder:default=info
+	// LogLevel define the ClusterLink components log level.
+	LogLevel string `json:"logLevel,omitempty"`
+
+	// +kubebuilder:default="ghcr.io/clusterlink-net"
+	// ContainerRegistry is the container registry to pull the ClusterLink project images.
+	ContainerRegistry string `json:"containerRegistry,omitempty"`
+	// +kubebuilder:default="latest"
+	// ImageTag represents the version of the ClusterLink project images.
+	ImageTag string `json:"imageTag,omitempty"`
+}
```

Comment:
future: add a destination namespace (for now it's fixed to `clusterlink-system`) but may want to have user control over that.
You can assume the namespace must exist before (e.g., created by the user when/before the operator is deployed) and fail if it does not.